sdk/identity/azure-identity/src/chained_token_credential.cpp

```diff
@@ -4,69 +4,138 @@
 #include "azure/identity/chained_token_credential.hpp"
 
 #include "azure/core/internal/diagnostics/log.hpp"
+#include <azure/core/azure_assert.hpp>
 
 #include <utility>
 
 using namespace Azure::Identity;
 using namespace Azure::Core::Credentials;
 using Azure::Core::Context;
+using Azure::Core::Diagnostics::Logger;
+using Azure::Core::Diagnostics::_internal::Log;
+
+namespace {
+std::string const IdentityPrefix = "Identity: ";
+}
 
 ChainedTokenCredential::ChainedTokenCredential(ChainedTokenCredential::Sources sources)
-    : m_sources(std::move(sources))
+    : ChainedTokenCredential(sources, {}, {})
 {
 }
 
-ChainedTokenCredential::~ChainedTokenCredential() = default;
-
-AccessToken ChainedTokenCredential::GetToken(
-    TokenRequestContext const& tokenRequestContext,
-    Context const& context) const
+ChainedTokenCredential::ChainedTokenCredential(
+    ChainedTokenCredential::Sources sources,
+    std::string const& enclosingCredential,
+    std::vector<std::string> sourceDescriptions)
+    : m_sources(std::move(sources)), m_sourceDescriptions(std::move(sourceDescriptions))
 {
-  using Azure::Core::Diagnostics::Logger;
-  using Azure::Core::Diagnostics::_internal::Log;
+  // LCOV_EXCL_START
+  AZURE_ASSERT(m_sourceDescriptions.empty() || m_sourceDescriptions.size() == m_sources.size());
```

Comment:
Sources should contain all information about the source including the friendly name we log.  Eliminates the need for this check.

Reply:
That will break compatibility. The only publically available constructor is with vector of credentials. It does not have a way to figure out credential types, other than via RTTI.

Reply:
Ah, I see, you mean that TokenCredential should have std::string Name(). I'll open an issue for that.

Reply:
#4418